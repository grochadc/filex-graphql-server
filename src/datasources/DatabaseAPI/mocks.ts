import { ParameterizedQuery as PQ } from "pg-promise";
import { students } from "datasources/StudentsAPI/mocks";
import {
  SELECT_STUDENT,
  SELECT_STUDENT_RESERVATION,
  SELECT_NEW_RESERVATION,
  SELECT_OPTIONS,
  DELETE_RESERVATIONS,
  SELECT_TEACHER_INFO,
  SELECT_TEACHER_RESERVATIONS,
  SELECT_SINGLE_OPTION,
  INSERT_RESERVATION,
  GET_TEACHER_RESERVATIONS,
  SELECT_TEACHERS,
  SIMPLE_TEACHER_INFO,
  DELETE_TEACHER_RESERVATIONS
} from "./index";

const teacher_reservations = [
  {
    workshop_id: 2,
    workshop_name: "Conversation",
    option_id: 2,
    codigo: "1234509876",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    nivel: 4,
    grupo: "E4-1"
  }
];

const reservations = [
  {
    id: 1,
    student_id: 2,
    option_id: 2,
    day: "Lunes",
    time: "1300 - 14:00",
    teacher_id: 1,
    teacher_name: "Fulanito",
    workshop_id: "workshop_id1",
    workshop_name: "Conversation",
    url: "http://url1"
  }
];
const teachersOptions = [
  {
    teacher_id: 1,
    options: [
      {
        day: "Miercoles",
        id: 4,
        name: "Gonzalo",
        option_id: 7,
        option_ids: [7, 8, 22],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/ukv-avjt-qiy",
        workshop_id: 3,
        workshop_name: "Conversation"
      },
      {
        day: "Jueves",
        id: 4,
        name: "Gonzalo",
        option_id: 8,
        option_ids: [7, 8, 22],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/ukv-avjt-qiy",
        workshop_id: 3,
        workshop_name: "Conversation"
      },
      {
        day: "Martes",
        id: 4,
        name: "Gonzalo",
        option_id: 22,
        option_ids: [7, 8, 22],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/kux-jhii-eyi",
        workshop_id: 4,
        workshop_name: "Tutoring"
      }
    ]
  },
  {
    teacher_id: 2,
    options: [
      {
        day: "Lunes",
        id: 3,
        name: "Sutanito",
        option_id: 5,
        option_ids: [5, 6, 21],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/sutanitolink1",
        workshop_id: 3,
        workshop_name: "Conversation"
      },
      {
        day: "Martes",
        id: 3,
        name: "Sutanito",
        option_id: 6,
        option_ids: [5, 6, 21],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/sutanitolink2",
        workshop_id: 3,
        workshop_name: "Conversation"
      },
      {
        day: "Viernes",
        id: 3,
        name: "Sutanito",
        option_id: 21,
        option_ids: [5, 6, 21],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/sutanitolink3",
        workshop_id: 4,
        workshop_name: "Tutoring"
      }
    ]
  },
  {
    teacher_id: 4,
    options: [
      {
        day: "Miercoles",
        id: 4,
        name: "Gonzalo",
        option_id: 7,
        option_ids: [7, 8, 22],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/ukv-avjt-qiy",
        workshop_id: 3,
        workshop_name: "Conversation"
      },
      {
        day: "Jueves",
        id: 4,
        name: "Gonzalo",
        option_id: 8,
        option_ids: [7, 8, 22],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/ukv-avjt-qiy",
        workshop_id: 3,
        workshop_name: "Conversation"
      },
      {
        day: "Martes",
        id: 4,
        name: "Gonzalo",
        option_id: 22,
        option_ids: [7, 8, 22],
        time: "13:00 - 14:00",
        url: "https://meet.google.com/kux-jhii-eyi",
        workshop_id: 4,
        workshop_name: "Tutoring"
      }
    ]
  }
];

type Mocks = {
  db: {
    [method: string]: jest.Mock<Promise<any>>;
  };
};
type MyCallback = (args: PQ) => any;
const promisedMock = (callback: MyCallback): jest.Mock<Promise<any>> => {
  return jest.fn(args => {
    return Promise.resolve(callback(args));
  });
};
const mocks: Mocks = {
  db: {
    manyOrNone: promisedMock(args => {
      if (args.text === GET_TEACHER_RESERVATIONS) {
        const option_id = args.values[0];
        return getTeacherReservationsByOptionId(option_id);
      }
      console.log(
        "no more mocked responses for manyOrNone returning null",
        args
      );
      return null;
    }),
    one: promisedMock(args => {
      if (args.text === INSERT_RESERVATION) return Promise.resolve({ id: 1 });
      if (args.text === SELECT_NEW_RESERVATION)
        return {
          id: 1,
          student_id: 2,
          option_id: 2,
          day: "Lunes",
          time: "1300 - 14:00",
          teacher_id: "teacher_id1",
          teacher_name: "Fulanito",
          workshop_id: "workshop_id1",
          workshop_name: "Conversation",
          url: "http://url1"
        };
      console.log("called one", args);
    }),
    none: promisedMock(args => {
      console.log("called none", args);
      return null;
    }),
    oneOrNone: promisedMock(args => {
      if (args.text === SELECT_STUDENT_RESERVATION) {
        const result = reservations.find(
          reservation => reservation.student_id === args.values[0]
        );
        return result;
      }
      console.log("no more mocked responses for any", args);
    }),
    many: promisedMock(args => {
      //@ts-ignore
      if (args === SELECT_TEACHERS) {
        return [
          { id: 1, name: "Fulanito", option_ids: [1, 2] },
          { id: 2, name: "Menganito", option_ids: [3, 4] }
        ];
      }
      //@ts-ignore
      if (args === SELECT_OPTIONS) {
        return [
          {
            id: 1,
            workshop_name: "Conversation",
            day: "Lunes",
            time: "1300 - 14:00",
            teacher_name: "Fulanito",
            url: "http://url1",
            teacher_id: 1,
            workshop_id: 1,
            workshop_description:
              "This is the conversation workshop description.",
            workshop_levels: [1, 2, 3, 4, 5, 6],
            reservations: 0
          },
          {
            id: 2,
            workshop_name: "Conversation",
            day: "Martes",
            time: "13:00 - 14:00",
            teacher_name: "Fulanito",
            url: "http://url2",
            teacher_id: 1,
            workshop_id: 1,
            workshop_description:
              "This is the conversation workshop description.",
            workshop_levels: [1, 2, 3, 4, 5, 6],
            reservations: 0
          }
        ];
      }
      if (args.text === SIMPLE_TEACHER_INFO) {
        const teacher_id = args.values[0];
        const options = teachersOptions.find(
          teacherOptions => teacherOptions.teacher_id === teacher_id
        ).options;
        if (options === undefined) {
          throw new Error("Couldn't find a teacher with id " + teacher_id);
        }
        return options;
      }
      console.log("no more mocked responses for many", args);
    })
  }
};

function getTeacherReservationsByOptionId(option_id: number) {
  const teacher_reservations = [
    {
      option_id: 7,
      reservations: [
        {
          workshop_id: 2,
          workshop_name: "Conversation",
          option_id: 2,
          codigo: "1234509876",
          nombre: "Benito Antonio",
          apellido_paterno: "Martinez",
          apellido_materno: "Ocasio",
          nivel: 4,
          grupo: "E4-1"
        }
      ]
    },
    {
      option_id: 8,
      reservations: [
        {
          workshop_id: 2,
          workshop_name: "Conversation",
          option_id: 2,
          codigo: "1234509876",
          nombre: "Benito Antonio",
          apellido_paterno: "Martinez",
          apellido_materno: "Ocasio",
          nivel: 4,
          grupo: "E4-1"
        }
      ]
    },
    {
      option_id: 22,
      reservations: [
        {
          workshop_id: 2,
          workshop_name: "Conversation",
          option_id: 2,
          codigo: "1234509876",
          nombre: "Benito Antonio",
          apellido_paterno: "Martinez",
          apellido_materno: "Ocasio",
          nivel: 4,
          grupo: "E4-1"
        }
      ]
    }
  ];
  const result = teacher_reservations.find(
    option => option.option_id === option_id
  );
  return result ? result.reservations : null;
}

export default mocks;
