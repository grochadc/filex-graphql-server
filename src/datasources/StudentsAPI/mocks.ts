import { ParameterizedQuery as PQ } from "pg-promise";
import { ADD_STUDENT, GET_STUDENT } from "./index";

export const applicants = [
  {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado",
    ciclo_ingreso: "2021A",
    telefono: "1234567890",
    email: "bad@bunny.pr",
    externo: false,
  },
  {
    id: 2,
    codigo: "0987654321",
    nombre: "Alberto",
    apellido_paterno: "Aguilera",
    apellido_materno: "Valadez",
    genero: "M",
    carrera: "Abogado",
    ciclo_ingreso: "2021A",
    telefono: "1234567890",
    email: "juanga@elnoanoa.mx",
    externo: false,
  },
];

export const students = [
  {
    id: 1,
    applicantId: 1,
    ciclo_actual: "2022B",
    codigo: "1234567890",
    nivel: 4, 
    curso: "en",
    groupId: 1,
    desertor: false,
    situacion: null,
  },
  {
    id: 2,
    applicantId: 2,
    ciclo_actual: "2022B",
    codigo: "0987654321",
    nivel: 3, 
    curso: "en",
    groupId: 2,
    desertor: false,
    situacion: null,
  }
]

const groups = [
  {
    id: 1,
    ciclo: "2022B",
    name: "E4-1",
    time: "07:00-08:00",
    aula: "G5",
    teacherId: 1,
    nivel: 4,
    registrados: 1,
    course: "en"
  },
  {
    id: 2,
    ciclo: "2022B",
    name: "E3-1",
    time: "07:00-08:00",
    aula: "G2",
    teacherId: 2,
    nivel: 3,
    registrados: 1,
    course: "en"
  }
];


export const studentss = [
  {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado",
    ciclo: "2021A",
    telefono: "1234567890",
    email: "bad@bunny.pr",
    nivel: "4",
    curso: "en",
    grupo: "E4-1",
    externo: false,
  },
  {
    id: 2,
    codigo: "0987654321",
    nombre: "Alberto",
    apellido_paterno: "Aguilera",
    apellido_materno: "Valadez",
    nivel: "1",
    grupo: "E1-1",
    genero: "M",
    carrera: "Abogado",
    ciclo: "2021A",
    telefono: "1234567890",
    email: "juanga@elnoanoa.mx",
    curso: "en",
    externo: false,
  },
];

const mocks = {
  db: {
    one: jest.fn((pq: PQ) => {
      if (pq.text === ADD_STUDENT) return Promise.resolve({ id: 1 });
    }),
    none: jest.fn((pq) => {
      return Promise.resolve();
    }),
    oneOrNone: jest.fn((pq) => {
      if (pq.text === GET_STUDENT) {
        const student = students.find(
          (student) => student.codigo === pq.values[0]
        );
        return Promise.resolve(student ? student : null);
      }
    }),
    any: jest.fn((pq) => {
      if (pq.text === GET_STUDENT) {
        const student = students.find(
          (student) => student.codigo === pq.values[0]
        );
        return Promise.resolve(student ? student : null);
      }
    }),
  },
};

export default mocks;
