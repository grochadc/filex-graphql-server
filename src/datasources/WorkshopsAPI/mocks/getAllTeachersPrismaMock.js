module.exports = [
  {
    "id": 1,
    "nombre": "Gonzalo Rocha",
    "options": [
      {
        "id": 1,
        "active": true,
        "day": "Lunes",
        "time": "13:00",
        "teacher_id": 1,
        "workshop_id": 1,
        "url": null,
        "place": "Centro de Aprendizaje Global",
        "workshop": {
          "id": 1,
          "name": "Conversation",
          "description": "Conversation Club",
          "levels": [
            1,
            2,
            3,
            4,
            5,
            6
          ]
        }
      }
    ]
  }
];