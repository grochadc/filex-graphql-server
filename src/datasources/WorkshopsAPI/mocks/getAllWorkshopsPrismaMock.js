module.exports = [
  {
    "id": 1,
    "name": "Conversation",
    "description": "Conversation Club",
    "levels": [
      1,
      2,
      3,
      4,
      5,
      6
    ],
    "options": [
      {
        "id": 1,
        "active": true,
        "day": "Lunes",
        "time": "13:00",
        "teacher_id": 1,
        "workshop_id": 1,
        "url": null,
        "place": "Centro de Aprendizaje Global",
        "teacher": {
          "id": 1,
          "nombre": "Gonzalo Rocha"
        }
      }
    ]
  }
];