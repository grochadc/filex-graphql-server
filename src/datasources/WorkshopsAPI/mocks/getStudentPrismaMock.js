module.exports = {
  "id": 3,
  "applicantId": 1,
  "ciclo_actual": "2022B",
  "codigo": "1234567890",
  "nivel": 4,
  "curso": "en",
  "situacion": null,
  "groupId": 2,
  "desertor": false,
  "applicant": {
    "id": 1,
    "codigo": "1234567890",
    "nombre": "Benito Antonio",
    "apellido_paterno": "Martinez",
    "apellido_materno": "Ocasio",
    "genero": "M",
    "carrera": "Abogado (DECH)",
    "ciclo_ingreso": "2022A",
    "telefono": "0001234567",
    "email": "bad@bunny.pr",
    "institutional_email": "benito.martinez@cusur.udg.mx",
    "externo": false
  },
  "groupObject": {
    "ciclo": "2022B",
    "name": "E4-1",
    "time": "08:00-09:00",
    "aula": "F4",
    "teacherId": 1,
    "id": 2,
    "nivel": 4,
    "registrados": 30,
    "course": "en"
  }
};