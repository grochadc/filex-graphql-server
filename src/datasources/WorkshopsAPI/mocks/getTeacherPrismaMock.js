module.exports = {
  "id": 1,
  "nombre": "Gonzalo Rocha"
};