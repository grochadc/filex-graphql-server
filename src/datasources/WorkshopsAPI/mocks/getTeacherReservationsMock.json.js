module.exports = [
  {
    "id": 1,
    "create_time": "2022-09-20T00:00:00.000Z",
    "student_id": 2,
    "option_id": 1,
    "attended": false,
    "option": {
      "id": 1,
      "active": true,
      "day": "Lunes",
      "time": "13:00",
      "teacher_id": 1,
      "workshop_id": 1,
      "url": null,
      "place": "Centro de Aprendizaje Global"
    },
    "student": {
      "id": 2,
      "applicantId": 2,
      "ciclo_actual": "2022B",
      "codigo": "0987654321",
      "nivel": 4,
      "curso": "en",
      "situacion": null,
      "groupId": 2,
      "desertor": false,
      "applicant": {
        "id": 2,
        "codigo": "0987654321",
        "nombre": "Alberto",
        "apellido_paterno": "Aguilera",
        "apellido_materno": "Valadez",
        "genero": "M",
        "carrera": "Agrobiotecnologia (AGB)",
        "ciclo_ingreso": "2022B",
        "telefono": "0001234567",
        "email": "juanga@elnoanoa.mx",
        "institutional_email": "alberto.aguilera@alumnos.udg.mx",
        "externo": false
      },
      "groupObject": {
        "ciclo": "2022B",
        "name": "E4-1",
        "time": "08:00-09:00",
        "aula": "F4",
        "teacherId": 1,
        "id": 2,
        "nivel": 4,
        "registrados": 30,
        "course": "en"
      }
    }
  },
  {
    "id": 2,
    "create_time": "2022-09-20T00:00:00.000Z",
    "student_id": 3,
    "option_id": 1,
    "attended": false,
    "option": {
      "id": 1,
      "active": true,
      "day": "Lunes",
      "time": "13:00",
      "teacher_id": 1,
      "workshop_id": 1,
      "url": null,
      "place": "Centro de Aprendizaje Global"
    },
    "student": {
      "id": 3,
      "applicantId": 1,
      "ciclo_actual": "2022B",
      "codigo": "1234567890",
      "nivel": 4,
      "curso": "en",
      "situacion": null,
      "groupId": 2,
      "desertor": false,
      "applicant": {
        "id": 1,
        "codigo": "1234567890",
        "nombre": "Benito Antonio",
        "apellido_paterno": "Martinez",
        "apellido_materno": "Ocasio",
        "genero": "M",
        "carrera": "Abogado (DECH)",
        "ciclo_ingreso": "2022A",
        "telefono": "0001234567",
        "email": "bad@bunny.pr",
        "institutional_email": "benito.martinez@cusur.udg.mx",
        "externo": false
      },
      "groupObject": {
        "ciclo": "2022B",
        "name": "E4-1",
        "time": "08:00-09:00",
        "aula": "F4",
        "teacherId": 1,
        "id": 2,
        "nivel": 4,
        "registrados": 30,
        "course": "en"
      }
    }
  }
];