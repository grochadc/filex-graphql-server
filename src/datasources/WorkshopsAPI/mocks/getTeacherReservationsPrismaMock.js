module.exports = [
  {
    "id": 1,
    "create_time": "2022-09-22T00:00:00.000Z",
    "student_id": 3,
    "option_id": 1,
    "attended": false,
    "option": {
      "id": 1,
      "active": true,
      "day": "Lunes",
      "time": "13:00",
      "teacher_id": 1,
      "workshop_id": 1,
      "url": null,
      "place": "Centro de Aprendizaje Global"
    },
    "student": {
      "id": 3,
      "applicantId": 1,
      "ciclo_actual": "2022B",
      "codigo": "1234567890",
      "nivel": 4,
      "curso": "en",
      "situacion": null,
      "groupId": 2,
      "desertor": false,
      "applicant": {
        "id": 1,
        "codigo": "1234567890",
        "nombre": "Benito Antonio",
        "apellido_paterno": "Martinez",
        "apellido_materno": "Ocasio",
        "genero": "M",
        "carrera": "Abogado (DECH)",
        "ciclo_ingreso": "2022A",
        "telefono": "0001234567",
        "email": "bad@bunny.pr",
        "institutional_email": "benito.martinez@cusur.udg.mx",
        "externo": false
      },
      "groupObject": {
        "ciclo": "2022B",
        "name": "E4-1",
        "time": "08:00-09:00",
        "aula": "F4",
        "teacherId": 1,
        "id": 2,
        "nivel": 4,
        "registrados": 30,
        "course": "en"
      }
    }
  }
];