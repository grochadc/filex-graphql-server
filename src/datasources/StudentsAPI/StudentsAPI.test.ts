import { StudentsAPI } from "../index";
import mocks, { studentss, students, applicants } from "./mocks";
import { prismaMock } from "./testutils/singleton";

const api = new StudentsAPI(mocks.db, prismaMock);

test("gets a student", async () => {
  const mock = jest.spyOn(prismaMock.student, "findFirst");
  mock.mockResolvedValueOnce({
    ...students[0],
    applicant: applicants[0]
  } as any);

  const result = await api.getStudent("1234567890", "2022B");
  expect(prismaMock.student.findFirst).toHaveBeenCalledWith({
    include: {
      applicant: true,
      groupObject: true
    },
    where: {
      ciclo_actual: "2022B",
      codigo: "1234567890"
    }
  });
  expect(result).toMatchSnapshot();
});

test("gets an inexistent student", async () => {
  const result = api.getStudent("1234509876", "2022B");
  await expect(result).rejects.toThrow();
});
test("adds a student", async () => {
  const result = await api.addStudent(studentss[0]);
  expect(result).toMatchSnapshot();
});

test("edits a student", async () => {
  const result = await api.editStudent(
    "0987654321",
    { carrera: "Periodismo" },
    "2022B"
  );
  expect(result).toMatchSnapshot();
});
