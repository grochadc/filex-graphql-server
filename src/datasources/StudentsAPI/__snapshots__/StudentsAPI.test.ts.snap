// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds a student 1`] = `
{
  "apellido_materno": "Ocasio",
  "apellido_paterno": "Martinez",
  "carrera": "Abogado",
  "ciclo": "2021A",
  "codigo": "1234567890",
  "curso": "en",
  "email": "bad@bunny.pr",
  "externo": false,
  "genero": "M",
  "grupo": "E4-1",
  "id": 1,
  "nivel": "4",
  "nombre": "Benito Antonio",
  "telefono": "1234567890",
}
`;

exports[`edits a student 1`] = `
Object {
  "apellido_materno": "Valadez",
  "apellido_paterno": "Aguilera",
  "carrera": "Periodismo",
  "ciclo": "2021A",
  "codigo": "0987654321",
  "curso": "en",
  "email": "juanga@elnoanoa.mx",
  "externo": false,
  "genero": "M",
  "grupo": "E1-1",
  "id": 2,
  "nivel": "1",
  "nombre": "Alberto",
  "telefono": "1234567890",
}
`;

exports[`gets a student 1`] = `
Object {
  "apellido_materno": "Ocasio",
  "apellido_paterno": "Martinez",
  "applicant": Object {
    "apellido_materno": "Ocasio",
    "apellido_paterno": "Martinez",
    "carrera": "Abogado",
    "ciclo_ingreso": "2021A",
    "codigo": "1234567890",
    "email": "bad@bunny.pr",
    "externo": false,
    "genero": "M",
    "id": 1,
    "nombre": "Benito Antonio",
    "telefono": "1234567890",
  },
  "applicantId": 1,
  "carrera": "Abogado",
  "ciclo_actual": "2022B",
  "ciclo_ingreso": "2021A",
  "codigo": "1234567890",
  "curso": "en",
  "desertor": false,
  "email": "bad@bunny.pr",
  "externo": false,
  "genero": "M",
  "groupId": 1,
  "id": 1,
  "nivel": 4,
  "nombre": "Benito Antonio",
  "situacion": null,
  "telefono": "1234567890",
}
`;
